# LOGISTIC REGRESSION

## CON FILTRO <40

In [26]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform
from sklearn.linear_model import LogisticRegression
log_reg=LogisticRegression(max_iter=10000,random_state=18)
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import pickle

In [27]:
with open("Data/train_data_40.pkl",'rb') as f:
    over_train_data_40 = pickle.load(f)

In [28]:
train_labels_40 = over_train_data_40['MVP']
train_data_40= over_train_data_40.drop('MVP', axis = 1)

In [29]:
C = uniform(loc=0, scale=1)
solver = ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga']
scoring = ['precision', 'f1', 'roc_auc']

rs_parameters_40 = {'C': C,
                 'solver' : solver,
                 }

In [11]:
#cross validation usando solo il train RandomSearchCV
rs_40=RandomizedSearchCV(log_reg, rs_parameters_40, scoring=scoring, refit="f1", n_jobs=-1, n_iter=250, random_state=756)
rs_lr_40 = rs_40.fit(train_data_40, train_labels_40)

In [13]:
print("The best score is",rs_lr_40.best_score_)
print("The best parameters are", rs_lr_40.best_params_)


The best parameters are {'C': 0.9442714194028784, 'solver': 'sag'}


Output Orfeo:

The best score is 0.9967138225402532
The best parameters are {'C': 0.9442714194028784, 'solver': 'sag'}


In [30]:
#fit su train con pramaetri trovati
log_reg_final_40=LogisticRegression(solver="sag",C=0.9443,class_weight="balanced", max_iter=10000,n_jobs=-1, random_state=13)
# log_reg_final_40=LogisticRegression(solver="linear",C=0.23927,class_weight="balanced", max_iter=10000,n_jobs=-1, random_state=13)

In [31]:
log_reg_final_40.fit(train_data_40, train_labels_40)

LogisticRegression(C=0.9443, class_weight='balanced', max_iter=10000, n_jobs=-1,
                   random_state=13, solver='sag')

In [32]:
with open("log_reg_40.pkl",'wb') as f:
    pickle.dump(log_reg_final_40,f)

In [33]:
with open("Data/test_data_40.pkl",'rb') as f:
    test_data_40 = pickle.load(f)

test_labels_40 = test_data_40["MVP"]
test_data_40=test_data_40.drop(["PLAYER_AGE", "PLAYER_ID", "SEASON_ID", "TEAM_ID", "MVP"], axis=1)

#predict su test
pred_lr_40 = log_reg_final_40.predict(test_data_40)

In [34]:
print(f"The accuracy of the Logistic Regressor is  {accuracy_score(test_labels_40,pred_lr_40)}")

The accuracy of the Logistic Regressor is  0.9924741298212606


In [35]:
#valutazione della prediction
confusion_matrix(test_labels_40,pred_lr_40)


array([[1051,    7],
       [   1,    4]])

In [36]:

print(classification_report(test_labels_40,pred_lr_40))

              precision    recall  f1-score   support

         0.0       1.00      0.99      1.00      1058
         1.0       0.36      0.80      0.50         5

    accuracy                           0.99      1063
   macro avg       0.68      0.90      0.75      1063
weighted avg       1.00      0.99      0.99      1063



## SENZA FILTRO <40

In [17]:
with open("Data/train_data.pkl",'rb') as f:
    over_train_data = pickle.load(f)

In [18]:
train_labels = over_train_data['MVP']
train_data= over_train_data.drop('MVP', axis = 1)

In [19]:
C = uniform(loc=0, scale=1)
solver = ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga']
scoring = ['precision', 'f1', 'roc_auc']

rs_parameters = {'C': C,
                 'solver' : solver,
                 }

In [5]:
#PROVARE CON N_ITER=SIZE(TRAIN_DATA_40) --> DOVREBBE ESSERE LEAVE ONE OUT CV
rs=RandomizedSearchCV(log_reg, rs_parameters, scoring=scoring, refit="f1", n_jobs=-1, n_iter=250, random_state=756)
rs_lr = rs.fit(train_data, train_labels)

In [6]:
#print("The best score is",rs_lr.best_score_)
print("The best parameters are", rs_lr.best_params_)

The best parameters are {'C': 0.9442714194028784, 'solver': 'sag'}


In [20]:
#fit su train con pramaetri trovati
log_reg_final=LogisticRegression(solver="sag",C=0.9443,class_weight="balanced", max_iter=10000,n_jobs=-1, random_state=13)

In [21]:
log_reg_final.fit(train_data, train_labels)

LogisticRegression(C=0.9443, class_weight='balanced', max_iter=10000, n_jobs=-1,
                   random_state=13, solver='sag')

In [22]:
with open("log_reg.pkl",'wb') as f:
    pickle.dump(log_reg_final,f)

In [23]:
with open("Data/test_data.pkl",'rb') as f:
    test_data = pickle.load(f)

test_labels = test_data["MVP"]
test_data=test_data.drop(["PLAYER_AGE", "PLAYER_ID", "SEASON_ID", "TEAM_ID", "MVP"], axis=1)

In [24]:
pred_lr = log_reg_final.predict(test_data)



print(f"The accuracy of the Logistic Regressor is  {accuracy_score(test_labels,pred_lr)}")

The accuracy of the Logistic Regressor is  0.9983543609434997


In [25]:
#valutazione della prediction
confusion_matrix(test_labels,pred_lr)

array([[1815,    3],
       [   0,    5]])

In [13]:
print(classification_report(test_labels,pred_lr))

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00      1818
         1.0       0.62      1.00      0.77         5

    accuracy                           1.00      1823
   macro avg       0.81      1.00      0.88      1823
weighted avg       1.00      1.00      1.00      1823



## Senza feature selection (pearson correlation)

In [3]:
with open("Data/train_data_no_pearson.pkl",'rb') as f:
    over_train_data = pickle.load(f)

In [4]:
train_labels = over_train_data['MVP']
train_data= over_train_data.drop('MVP', axis = 1)

In [5]:
C = uniform(loc=0, scale=1)
solver = ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga']
scoring = ['precision', 'f1', 'roc_auc']

rs_parameters = {'C': C,
                 'solver' : solver,
                 }

In [6]:
#PROVARE CON N_ITER=SIZE(TRAIN_DATA_40) --> DOVREBBE ESSERE LEAVE ONE OUT CV
rs=RandomizedSearchCV(log_reg, rs_parameters, scoring=scoring, refit="f1", n_jobs=-1, n_iter=250, random_state=756)
rs_lr = rs.fit(train_data, train_labels)

In [7]:
print("The best score is",rs_lr.best_score_)
print("The best parameters are", rs_lr.best_params_)

The best score is 0.9965778637772565
The best parameters are {'C': 0.9442714194028784, 'solver': 'sag'}


In [8]:
#fit su train con pramaetri trovati
log_reg_final=LogisticRegression(solver="sag",C=0.9443,class_weight="balanced", max_iter=10000,n_jobs=-1, random_state=13)

In [9]:
log_reg_final.fit(train_data, train_labels)

LogisticRegression(C=0.9443, class_weight='balanced', max_iter=10000, n_jobs=-1,
                   random_state=13, solver='sag')

In [10]:
with open("log_reg_no_pearson.pkl",'wb') as f:
    pickle.dump(log_reg_final,f)

In [13]:
with open("Data/test_data_no_pearson.pkl",'rb') as f:
    test_data = pickle.load(f)

test_labels = test_data["MVP"]
test_data=test_data.drop(["PLAYER_AGE", "PLAYER_ID", "SEASON_ID", "TEAM_ID", "MVP"], axis=1)

In [14]:
pred_lr = log_reg_final.predict(test_data)



print(f"The accuracy of the Logistic Regressor is  {accuracy_score(test_labels,pred_lr)}")

The accuracy of the Logistic Regressor is  0.9983543609434997


In [15]:
#valutazione della prediction
confusion_matrix(test_labels,pred_lr)

array([[1815,    3],
       [   0,    5]])

In [16]:
print(classification_report(test_labels,pred_lr))

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00      1818
         1.0       0.62      1.00      0.77         5

    accuracy                           1.00      1823
   macro avg       0.81      1.00      0.88      1823
weighted avg       1.00      1.00      1.00      1823

